flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/CatalogManager.java

```diff
@@ -697,6 +703,9 @@ private void dropTemporaryTableInternal(
         CatalogBaseTable catalogBaseTable = temporaryTables.get(objectIdentifier);
         if (filter.test(catalogBaseTable)) {
             temporaryTables.remove(objectIdentifier);
+            Optional<TemporaryOperationListener> listener =
+                    getTemporaryOperationListener(objectIdentifier);
+            listener.ifPresent(l -> l.onDropTemporaryTable(objectIdentifier.toObjectPath()));
```

Comment:
I think we should move this action before `temporaryTables.remove(objectIdentifier)`, otherwise, there might be data inconsistent if `onDropTemporaryTable` throws exception. We also put `onCreateTemporaryTable` before put into temporary map. 

The same to drop temporary function.

Reply:
Actually I intentionally put `onDropTemporaryTable` after `temporaryTables.remove(objectIdentifier)` here. My assumption is that `onDropTemporaryTable` is usually the best efforts to release some kind of resource associated with the temp table. And if that's the case, it's better to release the resource after the table has been dropped. WDYT?

Reply:
I think we should keep the behavior consistent for onCreate and onDrop, different connectors may have different needs on the listener, is there any connectors must do something before the temporary table is dropped? 

We can explicitly document the behavior in javadocs that all the methods are invoked before temporary object is registered/unregistered, and any thrown exception will prevent the registering/unregistering.

If we have any best effort logic, they can catch the exception instead of throwing it in the connectors's implementation.

Reply:
What do you think @JingsongLi ?

Reply:
Hi @wuchong , I want to avoid the situation when `onDropTemporaryTable` succeeds but "removing table" fails, in which case the table is still there but is likely to be unusable. But I also agree there may be connectors that want to do something before, rather than after, the table is dropped.
Maybe another option is to add interfaces both before and after the operation is performed? Something like this:
```
	Object preCreateTemporaryObject(ObjectType typeEnum, ObjectPath objectPath, Object object);

	void postCreateTemporaryObject(ObjectType typeEnum, ObjectPath objectPath, Object object);

	void preDropTemporaryObject(ObjectType typeEnum, ObjectPath objectPath);

	void postDropTemporaryObject(ObjectType typeEnum, ObjectPath objectPath);
```

Reply:
I also agree `onDropTemporaryTable succeeds but removing table fails` is meaningless. What I want to avoid is `removing table succeeds but onDropTemporaryTable fails`, that means users may not have chance to remove the temporary data.

Reply:
I thought more about this and since removing the table is unlikely fail compared to `onDropTemporaryTable`, I agree we should call `onDropTemporaryTable` before the table is removed.